src/main/java/org/truffleruby/cext/CExtNodes.java

```diff
@@ -1043,17 +1012,51 @@ public DynamicObject classNew(VirtualFrame frame, DynamicObject superclass) {
     @CoreMethod(names = "rb_tr_debug", onSingleton = true, rest = true)
     public abstract static class DebugNode extends CoreMethodArrayArgumentsNode {
 
+        @Child CallDispatchHeadNode toSCall;
+
         @TruffleBoundary
         @Specialization
         public Object debug(Object... objects) {
             if (objects.length > 1) {
                 System.err.printf("Printing %d values%n", objects.length);
             }
+
             for (Object object : objects) {
-                System.err.printf("%s @ %s: %s%n", object.getClass(), System.identityHashCode(object), object);
+                final String representation;
+
+                if (RubyGuards.isRubyString(object)) {
+                    final Rope rope = StringOperations.rope((DynamicObject) object);
+                    final byte[] bytes = rope.getBytes();
+                    final StringBuilder builder = new StringBuilder();
+
+                    for (int i = 0; i < bytes.length; i++) {
+                        if (i % 4 == 0 && i != 0 && i != bytes.length - 1) {
+                            builder.append(" ");
+                        }
+                        builder.append(String.format("%02x", bytes[i]));
+                    }
+
+                    representation = RopeOperations.decodeRope(rope) + " (" + ((Object) builder).toString() + ")";
+                } else if (RubyGuards.isRubyBasicObject(object)) {
+                    representation = object.toString() + " (" + StringOperations.getString(callToS(object)) + ")";
+                } else {
+                    representation = object.toString();
+                }
+
+                System.err.printf("%s @ %s: %s%n", object.getClass(), System.identityHashCode(object), representation);
             }
             return nil();
         }
+
+        private DynamicObject callToS(Object object) {
+            if (toSCall == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                toSCall = CallDispatchHeadNode.createOnSelf();
+            }
+
+            return (DynamicObject) toSCall.call(null, object, "to_s");
```

Comment:
Would `inspect` be better than `to_s`?

Reply:
I was afraid it could lead to too big outputs.